src/transform.rs

```diff
@@ -0,0 +1,142 @@
+use std::{error::Error, fmt};
+
+use crate::{Proj, ProjError};
+
+/// Transform a geometry using PROJ.
+pub trait Transform<T> {
+    type Output;
+
+    /// Transform a Geometry by mutating it in place.
+    ///
+    #[cfg_attr(feature = "geo-types", doc = r##"
+# Examples
+
+Transform a geometry using a PROJ string definition:
+
+```
+use geo_types;
+use proj::{Proj, Transform};
+ # use approx::assert_relative_eq;
+
+let mut point = geo_types::point!(x: -36.508f32, y: -54.2815f32);
+let proj = Proj::new("+proj=axisswap +order=2,1,3,4").expect("invalid proj string");
+ point.transform(&proj);
+
+assert_relative_eq!(
+    point,
+    geo_types::point!(x: -54.2815f32, y: -36.508f32)
+);
+```
+"##)]
+    fn transform(&mut self, proj: &Proj) -> Result<(), ProjError>;
+
+    /// Immutable flavor of [`Transform::transform`], which allocates a new geometry.
+    ///
+    #[cfg_attr(feature = "geo-types", doc = r##"
+# Examples
+
+Transform a geometry using a PROJ string definition:
+
+```
+use geo_types;
+use proj::{Proj, Transform};
+# use approx::assert_relative_eq;
+
+let point = geo_types::point!(x: -36.508f32, y: -54.2815f32);
+let proj = Proj::new("+proj=axisswap +order=2,1,3,4").expect("invalid proj string");
+
+assert_relative_eq!(
+    point.transformed(&proj).unwrap(),
+    geo_types::point!(x: -54.2815f32, y: -36.508f32)
+);
+
+// original `point` is untouched
+assert_relative_eq!(
+    point,
+    geo_types::point!(x: -36.508f32, y: -54.2815f32)
+);
+```
+"##)]
+    fn transformed(&self, proj: &Proj) -> Result<Self::Output, ProjError>;
```

Comment:
Should we add [`must_use`](https://rust-lang.github.io/rfcs/1940-must-use-functions.html) here ?